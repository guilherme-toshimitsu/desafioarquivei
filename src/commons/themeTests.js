import React from "react";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import * as Redux from "react-redux";
import thunk from "redux-thunk";
import theme from "./theme";
import store from "@store";

const renderWithTheme = (children) => {
  return render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
};

const renderWithThemeAndRedux = (children) => {
  return render(
    <ThemeProvider theme={theme}>
      <Redux.Provider store={store}>{children}</Redux.Provider>
    </ThemeProvider>
  );
};

export { renderWithTheme, renderWithThemeAndRedux };
