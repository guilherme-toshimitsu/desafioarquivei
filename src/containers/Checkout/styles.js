import { memo } from "react";
import styled from "styled-components";
import { breakpoints } from "@commons/theme";

const ButtonContainer = memo(styled.div`
  display: flex;
  justify-content: flex-end;
  ${breakpoints.sm} {
    justify-content: flex-start;
  }
`);

export { ButtonContainer };
