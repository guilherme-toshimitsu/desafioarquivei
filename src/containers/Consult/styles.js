import { memo } from "react";
import styled from "styled-components";

import { breakpoints, points } from "@commons/theme";

const ButtonContainer = styled.div`
  margin-top: 140px;
  display: flex;
  justify-content: flex-end;
  ${breakpoints.sm} {
    justify-content: flex-start;
  }
`;

export { ButtonContainer };
