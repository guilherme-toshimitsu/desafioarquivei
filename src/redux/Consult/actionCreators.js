import CONSTANTS from "./constants";
import Api from "@api";
import Router from "next/router";

const sendPriceAmmount = (price, ammount) => (dispatch) => {
  dispatch({
    type: CONSTANTS.CALCULATE_PRICE,
    price,
    consultsToBePurchased: ammount,
  });
};

const purchase = (data) => (dispatch) => {
  dispatch({
    type: CONSTANTS.CONSULT_PURCHASE,
  });
  Api.checkout
    .purchaseConsults(data)
    .then((res) => {
      dispatch({
        type: CONSTANTS.CONSULT_PURCHASE_SUCCESS,
      });
      alert("Success");
      Router.push("/");
    })
    .catch((err) => {
      dispatch({
        type: CONSTANTS.CONSULT_PURCHASE_FAIL,
      });
      alert("Error :", err);
    });
};

export default { sendPriceAmmount, purchase };
