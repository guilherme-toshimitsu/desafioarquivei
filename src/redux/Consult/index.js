export { default as actions } from "./actionCreators";
export { default } from "./reducer";
