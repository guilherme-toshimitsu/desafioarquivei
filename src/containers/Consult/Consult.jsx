import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import Router from "next/router";

import Button from "@components/Button";
import Card from "@components/Card";
import InsideContainer from "@components/InsideContainer";
import { actions } from "@store/Consult";
import { calculatePrice } from "@commons/utils";
import { InputRedux } from "@components/Form";
import { Title, Text } from "@components/Texts";

import { ButtonContainer } from "./styles";

const Consult = () => {
  // const { email } = useSelector((state) => state.userStore);
  const { consults, isLoading } = useSelector((state) => state.consultStore);
  const dispatch = useDispatch();
  const [toBePurchased, setToBePurchased] = useState(0);
  const [price, setPrice] = useState(0);

  const updateToBePurchased = (e) => {
    setToBePurchased(Number(e.target.value));
  };

  const checkPrice = () => {
    setPrice(calculatePrice(consults, toBePurchased));
  };

  const dispatchAmmountToBePurchased = () => {
    if (!isLoading) {
      dispatch(actions.sendPriceAmmount(price, toBePurchased));
      Router.push("/checkout");
    }
  };

  return (
    <InsideContainer>
      <Card>
        <Title>Consultas</Title>
        <Text>{`Consultas realizadas previamente: ${consults} `}</Text>

        <InputRedux
          type="number"
          label={"Quantidade de Consultas"}
          value={toBePurchased}
          input={{ onChange: updateToBePurchased }}
        />
        <Text>{`Preco: ${price.toFixed(2)} reais`}</Text>
        <ButtonContainer>
          <Button onClick={() => checkPrice()}>Calcular</Button>
          <Button
            onClick={() => dispatchAmmountToBePurchased()}
            disabled={!price || isLoading}
          >
            Comprar
          </Button>
        </ButtonContainer>
      </Card>
    </InsideContainer>
  );
};

export default Consult;
