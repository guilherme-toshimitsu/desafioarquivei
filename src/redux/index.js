import { createStore, applyMiddleware } from "redux";
import { Reducers } from "./reducers";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";

const store = createStore(
  Reducers,
  {},
  composeWithDevTools(applyMiddleware(thunk))
);

export default store;
