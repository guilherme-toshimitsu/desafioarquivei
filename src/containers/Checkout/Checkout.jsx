import React from "react";
import { useSelector, useDispatch } from "react-redux";
import Link from "next/link";
import Card from "@components/Card";
import InsideContainer from "@components/InsideContainer";
import PaymentForm from "./components/PaymentForm";
import { actions } from "@store/Consult";
import { Text, Title } from "@components/Texts";
import Loading from "@components/Loading";

const Checkout = () => {
  const { consultsToBePurchased, price, isLoading, user } = useSelector(
    (state) => state.consultStore
  );

  const dispatch = useDispatch();

  return (
    <InsideContainer>
      <Card>
        {price ? (
          <>
            {isLoading ? (
              <Loading />
            ) : (
              <>
                <Text data-testid={"true-loaded"}>
                  {`${consultsToBePurchased} consultas a serem compradas por ${price.toFixed(
                    2
                  )} reais`}
                </Text>
                <PaymentForm
                  initialValues={{ ...user }}
                  onSubmit={(data) => dispatch(actions.purchase(data))}
                />
              </>
            )}
          </>
        ) : (
          <>
            <Title>Compra invalida</Title>
            <Link href="/consult">Ir para Consultas</Link>
          </>
        )}
      </Card>
    </InsideContainer>
  );
};

export default Checkout;
