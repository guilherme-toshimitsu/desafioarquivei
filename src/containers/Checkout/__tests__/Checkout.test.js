import React, { useEffect } from "react";
import Checkout from "../";
import { renderWithThemeAndRedux, renderWithTheme } from "@commons/themeTests";
import { useSelector, useDispatch } from "react-redux";
import {
  fireEvent,
  waitForElement,
  cleanup,
  wait,
} from "@testing-library/react";
import envConfig from "@config";
import { actions } from "../../../redux/Consult";

describe("<Checkout />", () => {
  const oldConfig = envConfig;

  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();
  });

  afterEach(() => {
    envConfig.api = oldConfig.api;
  });

  it("creates a Checkout component without the load from Consult", () => {
    const wrapper = renderWithThemeAndRedux(<Checkout />);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("creates a Checkout component, after consults are loaded", async () => {
    envConfig.api.mock = true;

    const WrapperDispatch = () => {
      const dispatch = useDispatch();

      useEffect(() => {
        dispatch(actions.sendPriceAmmount(9, 100));
      }, []);

      return <></>;
    };

    const component = renderWithThemeAndRedux(
      <>
        <WrapperDispatch />
        <Checkout />
      </>
    );

    await waitForElement(() => component.getByTestId("true-loaded"));
    expect(component.asFragment()).toMatchSnapshot();
  });
});
