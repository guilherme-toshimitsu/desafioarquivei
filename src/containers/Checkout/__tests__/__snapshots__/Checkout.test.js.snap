// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkout /> creates a Checkout component without the load from Consult 1`] = `
<DocumentFragment>
  <div
    class="InsideContainer__Container-sc-1t900ay-0 cAJihm"
  >
    <div
      class="Card__StyledCard-sc-19fims7-0 HtJiy"
    >
      <p
        class="Texts__Title-pq9w5c-0 fiQxoW"
      >
        Compra invalida
      </p>
      <a
        href="/consult"
      >
        Ir para Consultas
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Checkout /> creates a Checkout component, after consults are loaded 1`] = `
<DocumentFragment>
  <div
    class="InsideContainer__Container-sc-1t900ay-0 cAJihm"
  >
    <div
      class="Card__StyledCard-sc-19fims7-0 HtJiy"
    >
      <p
        class="Texts__Text-pq9w5c-1 khPFWE"
        data-testid="true-loaded"
      >
        100 consultas a serem compradas por 9.00 reais
      </p>
      <form>
        <p
          class="Texts__Title-pq9w5c-0 fiQxoW"
        >
           Informacoes de Cartao
        </p>
        <div
          class="styles__InputContainer-sc-1ucodq7-0 fyGjJo"
        >
          <p
            class="styles__StyledLabel-sc-1ucodq7-3 eRPBMe"
          >
            CNPJ
          </p>
          <input
            class="styles__StyledInput-sc-1ucodq7-1 fJBjcv"
            name="CNPJ"
            placeholder=""
            type="text"
            value="___.___.___-__"
          />
        </div>
        <div
          class="styles__InputContainer-sc-1ucodq7-0 fyGjJo"
        >
          <p
            class="styles__StyledLabel-sc-1ucodq7-3 eRPBMe"
          >
            Owner
          </p>
          <input
            class="styles__StyledInput-sc-1ucodq7-1 fJBjcv"
            name="name"
            placeholder=""
            type="text"
            value=""
          />
        </div>
        <div
          class="styles__InputContainer-sc-1ucodq7-0 fyGjJo"
        >
          <p
            class="styles__StyledLabel-sc-1ucodq7-3 eRPBMe"
          >
            Card Number
          </p>
          <input
            class="styles__StyledInput-sc-1ucodq7-1 fJBjcv"
            name="cardnumber"
            placeholder=""
            type="text"
            value=""
          />
        </div>
        <div
          class="styles__InputContainer-sc-1ucodq7-0 fyGjJo"
        >
          <p
            class="styles__StyledLabel-sc-1ucodq7-3 eRPBMe"
          >
            CVV
          </p>
          <input
            class="styles__StyledInput-sc-1ucodq7-1 fJBjcv"
            name="cvv"
            placeholder=""
            type="text"
            value="___"
          />
        </div>
        <div
          class="styles__InputContainer-sc-1ucodq7-0 fyGjJo"
        >
          <p
            class="styles__StyledLabel-sc-1ucodq7-3 eRPBMe"
          >
            Card Expiration Date
          </p>
          <input
            class="styles__StyledInput-sc-1ucodq7-1 fJBjcv"
            name="cardexpiration"
            placeholder=""
            type="text"
            value="__/__/____"
          />
        </div>
        <div
          class="styles__ButtonContainer-ap13ju-0 hdiMdO"
        >
          <button
            class="Button__StyledButton-pzevnv-0 kodjjF"
            disabled=""
            type="submit"
          >
            Comprar
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
